### ISSUE TITLE
Not possible to deploy the devfile registry via CR on OpenShift 4.12

### ISSUE THREAD

**Issue by ibuziuk** [devfile/api#1092]
## Which area this feature is related to?

/kind bug

### Which area this bug is related to?

> /area registry

## What versions of software are you using?

latest from main

### Go project

**Operating System and version:**

OpenShift 4.12

**Go Pkg Version:**

### Node.js project

**Operating System and version:**

**Node.js version:**

**Yarn version:**

**Project.json:**

### Web browser

**Operating System and version:**

**Browser name and version:**

## Bug Summary

**Describe the bug:**

**To Reproduce:**

Install registry operator on 4.12 
create `devfile-registry` namespace and create CR in it 
based on `quay.io/devfile/devfile-index:next`

ERROR: pod can not be started:

```
Error creating: pods "devfile-registry-cd79f67d9-" is forbidden: unable to validate against any security context constraint: [provider "anyuid": Forbidden: not usable by user or serviceaccount, provider restricted-v2: .spec.securityContext.fsGroup: Invalid value: []int64{3001}: 3001 is not an allowed group, spec.containers[0].securityContext.runAsUser: Invalid value: 1001: must be in the ranges: [1000600000, 1000609999], spec.containers[1].securityContext.runAsUser: Invalid value: 1001: must be in the ranges: [1000600000, 1000609999], spec.containers[2].securityContext.runAsUser: Invalid value: 1001: must be in the ranges: [1000600000, 1000609999], provider "restricted": Forbidden: not usable by user or serviceaccount, provider "nonroot-v2": Forbidden: not usable by user or serviceaccount, provider "nonroot": Forbidden: not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount, provider "hostnetwork-v2": Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount, provider "privileged": Forbidden: not usable by user or serviceaccount]
```

## Additional context

![image](https://user-images.githubusercontent.com/1461122/229507676-1c6134db-1c2f-419a-a2b1-f6e740ef2ca1.png)
### Any workaround?

N / A

### Suggestion on how to fix the bug
related docs - https://devfile.io/docs/2.2.0/deploying-a-devfile-registry#deploying-a-devfile-registry-with-operator-lifecycle-manager

**Target Date: 05-02-2023**

**Comment by michael-valdron** [comment]
After further investigating we have found a few items to address:
1. Under no circumstances is the *default* namespace to used for the deployment target. This namespace does not respect any security contexts set and is not recommend for use with OpenShift deployments. To properly address this: 
    1. We will provide this requirement in our `README` instructions and/or our documentation
    2. Add validation in the webhooks to not allow the default namespace to be used for deployment
2. Our security contexts for the registry operator needs some adjusting to match the setups for our other templates, such as the helm chart. This is due to needing specific user/group definitions to allow a storage enabled volume for OCI registry to work, *which is enabled by default for the registry operator*. This feature is no longer needed for the devfile registry service so we will be disabling it by default to address this fix then further addressing the retirement of it across all deployment mechanisms in https://github.com/devfile/api/issues/1093.

**Comment by michael-valdron** [comment]
@ibuziuk In this thread, could you provide the linked issue(s) which are being blocked by this issue.

**Comment by michael-valdron** [comment]
Storage enabled volume now defaults to false fixing the security context issues: https://github.com/devfile/registry-operator/pull/40

Additional PRs coming in the next sprint for updating the documentation and providing webhook validation.

**Comment by ibuziuk** [comment]
just for the record, the issue happen in any namespace, not just default e.g. 
![image](https://user-images.githubusercontent.com/1461122/229770095-cf8fa293-bd85-40d6-9e74-3c714417d943.png)

**Comment by michael-valdron** [comment]
@ibuziuk 
> just for the record, the issue happen in any namespace, not just default e.g. ![image](https://user-images.githubusercontent.com/1461122/229770095-cf8fa293-bd85-40d6-9e74-3c714417d943.png)
> 
> > could you provide the linked issue(s) which are being blocked by this issue.

Thanks for catching and addressing this. We also caught this too during our deployment testing and disabling the storage enabled volume https://github.com/devfile/api/issues/1092#issuecomment-1495075473 seems to have fixed this during additional tests with OCP 4.12 environments.

> well, I put the blocker label since it was not possible to use the operator for deploying registries via CR on 4.11 and 4.12

This is understandable that any issue needing to use the registry operator will be blocked by this issue. With the blocker label, it is good to have what issue(s) are being blocked which triggered the creation of the issue in addition to any additional issues that others can report on in the thread, helps us keep track of what is going on.

*Side note*: Officially, we are only supporting OCP 4.12 at this point. This will added to our requirements in documentation as one of the TODOs for this issue.

**Comment by ibuziuk** [comment]
thanks, I added some details in the spike we had on the Eclipse Che end - https://github.com/eclipse/che/issues/22075#issuecomment-1496281368

**Comment by michael-valdron** [comment]
Work in progress has webhook validation which prints the following error message when trying to deploy a devfile registry to a default namespace:
```
Error from server (devfile registry deployment namespace should never be 'default'.): error when creating "registry.yaml": admission webhook "vdevfileregistry.kb.io" denied the request: devfile registry deployment namespace should never be 'default'.
```

**Comment by kim-tsao** [comment]
> Work in progress has webhook validation which prints the following error message when trying to deploy a devfile registry to a default namespace:
> 
> ```
> Error from server (devfile registry deployment namespace should never be 'default'.): error when creating "registry.yaml": admission webhook "vdevfileregistry.kb.io" denied the request: devfile registry deployment namespace should never be 'default'.
> ```

- @michael-valdron that looks good but can we avoid the duplicate error message `devfile registry deployment namespace should never be 'default'` (caused by error wrapping I think)

- Should we say something along the lines of `'default' namespace is forbidden for devfile registry deployment.  Retry the deployment with a non-default namespace` That way, users are guided on the corrective action to take.

**Comment by michael-valdron** [comment]
Updating to webhook validation test cases to take these changes into account.

**Comment by michael-valdron** [comment]
Webhook changes to validate the namespace is not default is now ready for review: https://github.com/devfile/registry-operator/pull/41

Additional PR is being worked on for documentation changes.

**Comment by michael-valdron** [comment]
Deployment requirements outline in [`README.md`](https://github.com/devfile/registry-operator/pull/42/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) is now ready for review: devfile/registry-operator#42

**Comment by michael-valdron** [comment]
Latest changes to the registry operator seems to have fixed the errors shown in this issue while using OCP 4.12 and Kubernetes 1.23.

@ibuziuk Can you confirm if this is fixed on your end?

**Comment by michael-valdron** [comment]
Additional time is needed for the review on this, target date has been updated. Confirmation of PR fixes and PR revisions to provide correct Kubernetes version requirements are still needed to close this issue.

**Comment by ibuziuk** [comment]
tested on 4.12 nightly and it seems to work :+1: 

![image](https://user-images.githubusercontent.com/1461122/233058940-01d34d2a-3472-4378-a3f0-c7ea94e6aae0.png)

**Comment by michael-valdron** [comment]
Now using issue #662 to track documentation additions related to the requirements these changes support.

Since it is confirmed that the changes completed in https://github.com/devfile/registry-operator/pull/40 and https://github.com/devfile/registry-operator/pull/41 have resolved this issue I will be closing it now.
